### PR TITLE
add support for preserving header key case

### PR DESCRIPTION
**Pull request by ramaraochavali**
Adds support for preserving header key case in HTTP 1.1

Fixes https://github.com/istio/istio/issues/32008

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[X] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[ ] Does not have any [user-facing](https://github.com/istio/istio/tree/master/releasenotes#when-to-add-release-notes) changes. This may include CLI changes, API changes, behavior changes, performance improvements, etc.

### CODE REVIEW

**Comment by ramaraochavali** [comment]
@su225 Sorry for hijacking your PR and API changes. But we have a high severity customer issue because of this and we need a fix immediately. Hence pushed this PR to fast forward the work. We were not able to do it via Envoy filter because of `proto.Merge` limitation on nested typed configs.

**Comment by su225** [comment]
@ramaraochavali - No problem :) I'll close mine.

**Comment by howardjohn** [comment]
I don't think we should do it but want to make sure we get some feedback on this. I think it may be fine but TOC has asked to be looped into any behavioral changes.

I think in the meantime you can use EnvoyFilter?
```
 apiVersion: networking.istio.io/v1alpha3
 kind: EnvoyFilter
 metadata:
   name: h2-control
 spec:
   configPatches:
   - applyTo: NETWORK_FILTER
     match:      
       listener:
         filterChain:
           filter:
             name: "envoy.filters.network.http_connection_manager"
     patch:
       operation: MERGE
       value:
         typed_config:
           "@type": "type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager"
           http_protocol_options:
             header_key_format:
               stateful_formatter:
                 name: preserve_case
                 typed_config:
                   "@type": type.googleapis.com/envoy.extensions.http.header_formatters.preserve_case.v3.PreserveCaseFormatterConfig
```

**Comment by howardjohn** [comment]
I would also like to know if there is any performance cost associated - the word "stateful" may be scaring me a bit

**Comment by ramaraochavali** [comment]
>  I think it may be fine but TOC has asked to be looped into any behavioral changes.

Fair enough.  We can default it to false if there is a concern for behavior change.

> I would also like to know if there is any performance cost associated - the word "stateful" may be scaring me a bit

There will be some memory overhead as it stores the keys 

https://github.com/envoyproxy/envoy/blob/main/source/extensions/http/header_formatters/preserve_case/preserve_case_formatter.h#L20 and minimal perf overhead to look at this. This totally depends on how many headers  a request has , but the general case may be fine.


> I think in the meantime you can use EnvoyFilter?

With that Envoy filter Istio sending this to Envoy and Envoy complaining about header_format missing - So I was wondering it is a proto.Merge issue.
```
http_protocol_options {
  header_key_format {
  }
}
```

**Comment by ramaraochavali** [comment]
Closing the loop on the filter discrepancy between what @howardjohn  is seeing on master and what I am seeing on `26614eba5d4485965d9ed26ddea8779aa591d86f` - the SHA I am testing does not have the updated go control plane and hence does not know about this type `github.com/envoyproxy/go-control-plane/envoy/extensions/http/header_formatters/preserve_case/v3`. 
`

File: releasenotes/notes/32008.yaml [hunk 1]
```diff
@@ -0,0 +1,9 @@
+apiVersion: release-notes/v2
+kind: feature
+area: traffic-management
+issue:
+- 32008
+releaseNotes:
+- |
+  **Added** support for preserving case of header key in HTTP 1.1 keys. This can be turned off by setting 
+  setting "PILOT_ENABLE_PRESERVE_HEADER_KEY_CASE"  to false.
```
Comment (shamsher31) [inline releasenotes/notes/32008.yaml:9]:
```suggestion
   `PILOT_ENABLE_PRESERVE_HEADER_KEY_CASE` to false.     
```

**Review by hzxuzhonghu — APPROVED** [review]

**Review by shamsher31 — APPROVED** [review]

**Comment by howardjohn** [comment]
I asked for this to not be merged until we look into this more. I already have it on the TOC agenda for today, depending on the outcome we may need to revert - just a heads up.

I talked with Rama and got the EnvoyFilter solution so I think they are unblocked anyhow (?)

**Comment by ramaraochavali** [comment]
Yes. If TOC feels change in behaviour is not acceptable, we can change the default to false?

**Comment by duderino** [comment]
My vote is to preserve the old behavior by default and expose this as an opt-in

**Comment by howardjohn** [comment]
We didn't get a chance to discuss with the full TOC but since the initial vote is -1 I think we may want to revert. There is an envoyfilter solution for those that require applications that do not follow the HTTP spec to use istio's HTTP support.

**Comment by ramaraochavali** [comment]
John when you mean revert - you meant default value right? I prefer having
this default off .

**Comment by nrjpoddar** [comment]
Agree about EnvoyFilter -> MeshConfig or first class API if it makes sense.

@ramaraochavali i didn’t get a satisfactory answer before but are there any security concerns for users who turn this on w.r.t header based AuthZ matches?

I’m still debating default on/off but keeping an ENV var is not optimal either way.

**Comment by ramaraochavali** [comment]
I am not against MeshConfig API.  However I do think this is an interim option (not a longer term option for many because of legacy clients migrating to newer stack) and also could not think of any downside of defaulting this to true in longer term in Istio (except for the behaviour change if any for existing users - which I really doubt because the newer http1.1 clients would any ways be sending headers in lower case already so there is no issue with them and it does not apply http2 at all). 

So I felt env var approach may be better here so that we can just remove it when we default to true for few releases and no one complains about it. However if every one thinks we should go via API - Let us do that. 

>  i didn’t get a satisfactory answer before but are there any security concerns for users who turn this on w.r.t header based AuthZ matches?

@nrjpoddar  I could not think of any.

**Comment by hzxuzhonghu** [comment]
This can be a temp solution before the  API lands on,  need to turn it off by default to make it safe.

**Comment by ramaraochavali** [comment]
If API is preferred and since we all agree that defaulting to false is better option, let us quickly converge on API https://github.com/istio/api/pull/2004. I can change the current implementation to use API instead of completely reverting it.

**Comment by howardjohn** [comment]
I am also seeing `ACK ERROR 764f7f92e61f.vm-namespace-12 Internal:Error adding/updating listener(s) 0.0.0.0_80: Proto constraint validation failed (HttpConnectionManagerValidationError.HttpProtocolOptions: ["embedded message failed validation"] | caused by Http1ProtocolOptionsValidationError.HeaderKeyFormat: ["embedded message failed validation"] | caused by field: "header_format", reason: is required): stat_prefix: "outbound_0.0.0.0_80"` as well

**Comment by ramaraochavali** [comment]
Is that VM running old version of proxy?

**Pull request merged**